### ISSUE TITLE
relative links do not work when including the file

### ISSUE THREAD

**Issue by jauco** [asciidoctor/asciidoctor#1903]
Showcase: https://github.com/jauco/asciidoctorbug

When including a file, relative links do not seem to be rewritten. So a link to ./some-file.txt from a subfolder now points to ./some-file.txt in the main folder.

Is this by design? If so: is there a workaround?

**Comment by koen-unifly** [comment]
FYI: the same happens with images.

If you have a relative path to an image (relative to the include-source file), after including it in the target file: the paths are not changed, resulting into broken image src-attributes. (Html 5 output).

**Comment by mojavelinux** [comment]
This is by design. It's an issue that comes up frequently. An include is a preprocessor directive. It does not know anything about the structure of the document, so it is not going to affect links. The best way to think of includes is that they all get resolved to create a single AsciiDoc document and that AsciiDoc document is sent to the parser (that's not exactly how it works, but it's effectively the same).

It's possible to include documents relative to other documents, which often misleads users into thinking that the document is aware of these directory changes.

For more background, see https://github.com/asciidoctor/asciidoctor-pdf/issues/93#issuecomment-98916327 (which references a bunch of issues).

**Comment by mojavelinux** [comment]
In other words, once the document is being parsed and converter, there is no longer any trace of the include files...they are intended strictly for concatenating sources.

However, you can redefine document attributes anywhere in the document to keep a relative reference to the "root", which you can use in the content. See the related issues for hints.

**Comment by DanielSWolf** [comment]
Has there been any progress? I love AsciiDoctor for single-file documents. But as soon as I split a document into multiple files, everything (links, images, attributes...) starts falling apart. Due to the "an include is just a macro" philosophy, no tool will show me a decent preview for the individual files, because they only make sense once they are included in the master file.

**Comment by mojavelinux** [comment]
@DanielSWolf I think what you are looking for is Antora. It's specifically designed to process and link multiple AsciiDoc files. And there is support for Antora in the AsciiDoc plugin for IntelliJ, so you will get the tooling assistance you are looking for.

> an include is just a macro

An include is not a macro, it is a preprocessor directive.

**Comment by DanielSWolf** [comment]
I'm aware of Antora, but I don't think it's a good match for my use case. I don't need a full-blown website with navigation and version selection. I need just a single HTML/PDF file as output. So I'm very happy with AsciiDoctorJS/PDF. I just wished they would make it easier to split one logical document into multiple files.

> An include is not a macro, it is a preprocessor directive.

Sorry, coming from C++, where both are the same.

**Comment by mojavelinux** [comment]
Many authors have been successful creating books that are split into multiple files, but that come together as a single file. There are rules that you need to learn and follow, but when you do it works very nicely (based on their feedback). Sometimes it helps to have a project template to guide you. One such project template is https://bootiful-asciidoctor.github.io/. Another is https://github.com/mraible/infoq-mini-book. Yet another is https://github.com/gregturn/my-tech-book-manuscript. There are others out there as well. I disagree things start falling apart when you use the include directive. They are designed a certain way and if you understand how to use them, they work exactly as they should.

**Comment by elextr** [comment]
@DanielSWolf if you are coming from a C++ perspective then it might be easier to understand if you consider that the Asciidoc include directive acts _exactly_ the same as the C++ `#include` directive.

What you are doing with your document is the equivalent of `#include`ing parts of the `.cpp` file from other files.  (Its not like `#include`ing headers, `.hpp` files)

If you have parts of a file that are `#include`ed in another file your IDE won't understand any context from where its included when you are editing the file because it just doesn't know where it might be used.

The same is true for Asciidoc previews, they don't know where a file is going to be included, so they can't consider any context from there.  You need to preview the constructed document, not disconnected parts.

**Comment by mojavelinux** [comment]
I also want to add/clarify that the AsciiDoc plugin for IntelliJ provides the necessary tooling for includes even when not using Antora (it detects which environment you are using). The core processor is not responsible for tooling, but the ecosystem around it is growing.

**Comment by DanielSWolf** [comment]
Thanks for your explanations and pointers, @mojavelinux and @elextr!

In hindsight, I see that my wording about things falling apart was unfortunate. I apologize if I offended you. That certainly wasn't my intention.

Now that I've got that off my chest: In [an old comment](https://github.com/asciidoctor/asciidoctor-pdf/issues/93#issuecomment-98916327), you mention that you might add "a block macro for subdocuments that has different behavior from include". Is this still a thing?

**Comment by mojavelinux** [comment]
No worries, I didn't take offense. I am firm on how AsciiDoc / Asciidoctor is characterized, that's all.

To follow the progress on the subdocument macro, see https://github.com/asciidoctor/asciidoctor/issues/3603 The proposal / idea is to introduce a subdocument macro into the language (likely by way of the specification process, at this point). However, since a block macro is an extension point, anyone can create such a macro at any time. There's no need to wait on core or the language.

There's no question a subdocument macro will be useful, but I think you can be perfectly successful using the include directive if you understand how it operates. The key is understanding that the include directive coalesces source files together. You can't think of links as relative to the document being included but rather relative to the document that is doing the including (i.e., the top document). In other words, they are really all one document in the end.

**Comment by mojavelinux** [comment]
...which is exactly what I said in my first comment:

> The best way to think of includes is that they all get resolved to create a single AsciiDoc document and that AsciiDoc document is sent to the parser (that's not exactly how it works, but it's effectively the same).

**Comment by Britz** [comment]
I have the same problem that the include directive breaks relative path in documents. 
I read through some related issues ([93](https://github.com/asciidoctor/asciidoctor-pdf/issues/93), [650](https://github.com/asciidoctor/asciidoctor/issues/650), [2608](https://github.com/asciidoctor/asciidoctor/issues/2608), [3603](https://github.com/asciidoctor/asciidoctor/issues/3603)) and this one seems to be the one with the newest contributions.

For short: Is there an official solution to that problem in the meantime?
I could neither find a ne subdocument directive (as mentioned in issue ) in or additional include parameters in the official docs. It is nearly 10 years old now.  

I fully understand the problem, that the include directive is a preprocessor step where the context is unknown and therefore you cannot simply inject variable definitions by default. The only exception to that is the level offset, since it is assumed, that you only specify that parameter when used at a section boundary (as mentioned in [issue 650](https://github.com/asciidoctor/asciidoctor/issues/650#issuecomment-433746305)). 

But why it is not possible to do just exactly the the same with attributes other than `leveloffset`.  
Assuming a context for a given attribute seems to be totally fine for me and it will still be backward compatible. Given an  attribute, e.g. `fixcontext` or `setdir`, the injection of asciidoc variable definitions coudl be triggerd before and after the included content, which could be used to set/unset the already mentioned `current_docdir` and `current_docfile` ([issue 650](https://github.com/asciidoctor/asciidoctor/issues/650)) or `indir` and `infile`  ([issue 2608](https://github.com/asciidoctor/asciidoctor/issues/2608)) variables , fixing `imagespath` or even overwriting the `docdir` and `docfile` variables.

As I know now from this research, there are a lot of work arounds out there, as using extensions – which most probably do not work with PDF, DOCX generators or Antora or break includes in code blocks or on other places – or making extensive use of variables in documents and around includes – which is highly error prone and hard to  understand for documentation contributors with less coding experience.

But these are still workarounds/hacks and not part of the asciidoc documentation.
